wagtail/query.py

```diff
@@ -489,85 +500,131 @@ def specific_iterator(qs, defer=False):
-            # Defer all specific fields
-            fields = [
-                field.attname for field in Page._meta.get_fields() if field.concrete
-            ]
-            pages = pages.only(*fields)
-        elif qs._defer_streamfields:
-            pages = pages.defer_streamfields()
-
-        pages_for_type = {page.pk: page for page in pages}
-        pages_by_type[content_type] = pages_for_type
-        missing_pks.extend(pk for pk in pks if pk not in pages_for_type)
-
-    # Fetch generic pages to supplement missing items
-    if missing_pks:
-        generic_pages = (
-            Page.objects.filter(pk__in=missing_pks)
-            .select_related("content_type")
-            .in_bulk()
-        )
-        warnings.warn(
-            "Specific versions of the following pages could not be found. "
-            "This is most likely because a database migration has removed "
-            "the relevant table or record since the page was created:\n{}".format(
-                [
-                    {"id": p.id, "title": p.title, "type": p.content_type}
-                    for p in generic_pages.values()
-                ]
-            ),
-            category=RuntimeWarning,
-        )
-    else:
-        generic_pages = {}
-
-    # Yield all pages in the order they occurred in the original query.
-    for pk, content_type in pks_and_types:
-        try:
-            page = pages_by_type[content_type][pk]
-        except KeyError:
-            page = generic_pages[pk]
+    while True:
+        if chunk_size:
+            slice_bottom = (chunk_start or 0) + ((chunk_number - 1) * chunk_size)
+            slice_top = (chunk_start or 0) + ((chunk_number * chunk_size))
+            if chunk_stop and slice_top > chunk_stop:
+                slice_top = chunk_stop
+            # Fetch one more than needed to determine if further chunks
+            # are available, without an extra database query
+            slice_top += 1
+            base_qs = qs[slice_bottom:slice_top]
+        else:
+            base_qs = qs
+
+        values = base_qs.values("pk", "content_type", *annotation_aliases)
+        annotations_by_pk = defaultdict(list)
         if annotation_aliases:
-            # Reapply annotations before returning
-            for annotation, value in annotations_by_pk.get(page.pk, {}).items():
-                setattr(page, annotation, value)
-        yield page
+            # Extract annotation results keyed by pk so we can reapply to fetched pages.
+            for data in values:
+                annotations_by_pk[data["pk"]] = {
+                    k: v for k, v in data.items() if k in annotation_aliases
+                }
+
+        pks_and_types = [[v["pk"], v["content_type"]] for v in values]
+
+        if chunk_size:
+            further_chunks_available = len(pks_and_types) > chunk_size
+            if further_chunks_available:
+                # Remove the 'over-fetched' item
+                pks_and_types.pop()
```

Comment:
The extra item is removed here, so the impact is not carried over to the specific queries